# Research Challenge

<div style="background-color: #f8d7da; border-left: 6px solid #ccc; margin: 20px; padding: 15px;">
    <strong>💡 Margaret Atwood:</strong> Every aspect of human technology has a dark side, including the bow and arrow.
</div>

## 🏅 Build your own model

It is time to go back to supervised machine learning problems.

You have been assigned one dataset from [MatBench](https://matbench.materialsproject.org) as introduced in the [Lecture slides](https://speakerdeck.com/aronwalsh/mlformaterials-challenge-25). You are free to choose and tune any machine-learning model, with any Python library, but it should be appropriate for the problem. For instance, [XGBoost](https://xgboost.readthedocs.io) could be a good starting starting point to build a regression model. You can refer back to earlier notebooks and repurpose code as needed.

You may reach the limits of computing processing power on Google Colab. Building a useful model with limited resources is a real-world skill. Using other free resources is allowed if you find an alternative service, as is running on your own computer. A model tracker such as [wandb](https://wandb.ai) could be helpful for advanced users. If you want to try a brute force approach, a library such as [Automatminer](https://hackingmaterials.lbl.gov/automatminer) may be of interest.

This notebook should be used for keeping a record of your model development, submission, and even your presentation. You are free to edit (add/remove/delete) or rearrange the cells as you see fit.

### Your details

In [23]:
import numpy as np

# Insert your values
Name = "Mark Zangwill" # Replace with your name
CID = 2198282 # Replace with your College ID (as a numeric value with no leading 0s)

# Set a random seed using the CID value
CID = int(CID)
np.random.seed(CID)

# Print the message
print("This is the work of " + Name + " [CID: " + str(CID) + "]\n")

# Define the available groups
groups = ['A', 'B', 'C', 'D', 'E']

# Select a group based on the seeded random state
challenge_group = np.random.choice(groups)

# Print the challenge code
print("Your challenge code is " + challenge_group)

This is the work of Mark Zangwill [CID: 2198282]

Your challenge code is C


## Problem statement
You have been assigned one dataset from the [list](https://matbench.materialsproject.org/Benchmark%20Info/matbench_v0.1/) on [MatBench](https://matbench.materialsproject.org). You should state what problem you are trying to solve and comment on the best-performing model in the benchmark.

In [24]:
# Spare cell




## Data preparation

Check the data distribution and apply appropriate pre-processing steps as required.

In [25]:
# Installation of libraries
!pip install matminer # Datasets and featurisation

In [26]:
# Get dataset info from matminer
from matminer.datasets import get_all_dataset_info
from matminer.datasets import load_dataset

# Detailed on https://hackingmaterials.lbl.gov/matminer/dataset_summary.html
# Uncomment the info line for your assigned challenge code

  # A
#info = get_all_dataset_info("matbench_dielectric")

  # B
#info = get_all_dataset_info("matbench_expt_gap")

  # C
info = get_all_dataset_info("matbench_expt_is_metal")

  # D
#info = get_all_dataset_info("matbench_glass")

  # E
#info = get_all_dataset_info("matbench_steels")

# Check the dataset information
print(info)

Dataset: matbench_expt_is_metal
Description: Matbench v0.1 test dataset for classifying metallicity from composition alone. Retrieved from Zhuo et al. supplementary information. Deduplicated according to composition, ensuring no conflicting reports were entered for any compositions (i.e., no reported compositions were both metal and nonmetal). For benchmarking w/ nested cross validation, the order of the dataset must be identical to the retrieved data; refer to the Automatminer/Matbench publication for more details.
Columns:
	composition: Chemical formula.
	is_metal: Target variable. 1 if is a metal, 0 if nonmetal.
Num Entries: 4921
Reference: Y. Zhuo, A. Masouri Tehrani, J. Brgoch (2018) Predicting the Band Gaps of Inorganic Solids by Machine Learning J. Phys. Chem. Lett. 2018, 9, 7, 1668-1673 
 https//:doi.org/10.1021/acs.jpclett.8b00124.
Bibtex citations: ["@Article{Dunn2020,\nauthor={Dunn, Alexander\nand Wang, Qi\nand Ganose, Alex\nand Dopp, Daniel\nand Jain, Anubhav},\ntitle={Benc

In [27]:
# Load your dataset into a pandas DataFrame
df = load_dataset("matbench_expt_is_metal")

print(df)

            composition  is_metal
0              Ag(AuS)2      True
1            Ag(W3Br7)2      True
2      Ag0.5Ge1Pb1.75S4     False
3     Ag0.5Ge1Pb1.75Se4     False
4                Ag2BBr      True
...                 ...       ...
4916             ZrTaN3     False
4917               ZrTe      True
4918             ZrTi2O      True
4919             ZrTiF6      True
4920               ZrW2      True

[4921 rows x 2 columns]


In [28]:
df.describe()

,composition,is_metal
count,4921,4921
unique,4921,2
top,ZrW2,False
freq,1,2470


Where middle column is the "composition" column, with all 4921 being unique. The last column beign the output, unique being two since outputs are either true or false, depending on whenther metal or non.

Choose relevant features, which may be based on composition or structure, depending on your problem. [matminer](https://hackingmaterials.lbl.gov/matminer/) is a good place to start.

###Materials Featurization

In [29]:
#download ElementEmbedings
!pip install elementembeddings

In [45]:
# Featurise all chemical compositions
from elementembeddings.composition import composition_featuriser

def featurise(dataframe):

  # Compute element embeddings using mean and max pooling
  mean_df = composition_featuriser(dataframe["composition"], embedding="magpie", stats=["mean"])
  max_df = composition_featuriser(dataframe["composition"], embedding="magpie", stats=["maxpool"])

  # Convert "is_metal" column to integer labels (0, 1)
  dataframe['is_metal'] = dataframe['is_metal'].astype(int)
  mean_df['is_metal'] = dataframe['is_metal']
  max_df['is_metal'] = dataframe['is_metal']

  # Define feature matrices and target variable
  cols_to_drop = ['is_metal', 'formula']

  X_mean = mean_df.drop(columns=cols_to_drop, errors='ignore').values
  X_max = max_df.drop(columns=cols_to_drop, errors='ignore').values
  y = dataframe['is_metal'].values  # Target variable

  return X_mean, X_max, y, mean_df, max_df

X_mean, X_max, y, mean_df, max_df = featurise(df)

print("Mean pooling features (first two rows, first 4 columns):")
print(mean_df.iloc[:2, :4])
print("\nMax pooling features (first two rows, first 4 columns):")
print(max_df.iloc[:2, :4])

Featurising compositions...


100%|██████████| 4921/4921 [00:12<00:00, 383.42it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 176104.65it/s]


Featurising compositions...


100%|██████████| 4921/4921 [00:12<00:00, 387.01it/s]


Computing feature vectors...


100%|██████████| 4921/4921 [00:00<00:00, 111128.78it/s]


Mean pooling features (first two rows, first 4 columns):
      formula  mean_Number  mean_MendeleevNumber  mean_AtomicWeight
0    Ag(AuS)2    47.400000                  74.6         113.186268
1  Ag(W3Br7)2    46.714286                  81.0         110.931629

Max pooling features (first two rows, first 4 columns):
      formula  maxpool_Number  maxpool_MendeleevNumber  maxpool_AtomicWeight
0    Ag(AuS)2            79.0                     88.0            196.966569
1  Ag(W3Br7)2            74.0                     95.0            183.840000


In [36]:
#print the number columns in the dataframes
print(mean_df.shape[1])
print(max_df.shape[1])
#print the number of rows
print(mean_df.shape[0])
print(max_df.shape[0])

24
24
4921
4921


## Model selection, testing and training

Define your model and justify your choice based on the problem and available data. You can look back at earlier notebooks and investigate other examples online including in [scikit-learn](https://scikit-learn.org).

In [40]:
#apply nested cross validation
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import  cross_val_score
from sklearn.metrics import roc_auc_score

decisionTreeClassifier = DecisionTreeClassifier(random_state= 42)
scores = cross_val_score(decisionTreeClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.81911451 0.82939767 0.83119061 0.80070644 0.84967363]
mean CV accuray 0.8260165722827468


In [41]:
from sklearn.ensemble import RandomForestClassifier

randomForestClassifier = RandomForestClassifier(random_state=42)
scores = cross_val_score(randomForestClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.94944012 0.95300132 0.97276915 0.9191089  0.95267083]
mean CV accuray 0.9493980637680008


In [42]:
# gradient boosted method
from sklearn.ensemble import GradientBoostingClassifier

gradientBoostingClassifier = GradientBoostingClassifier(random_state=42)
scores = cross_val_score(gradientBoostingClassifier, X_mean, y, cv=5, scoring='roc_auc')
print("ROC-AUC scores:", scores)
print("mean CV accuray", scores.mean())

ROC-AUC scores: [0.94248291 0.94673015 0.96578947 0.90432744 0.93843055]
mean CV accuray 0.9395521056283196


To conduct hyperparameter optimization, nested crossed validation is requred as to not leak data, giving optimisitic results. This involves having two loops; an inner and outer. The inner is where featurization is done as well as hyperparameter optimization. the outer does test.

Train, validate and test your model. Make sure to do proper data splits and to consider the hyperparamaters of your model.

<details>
<summary>Note on the ROC-AUC classification metric</summary>
There is one metric we didn't cover but is used in Matbench. In binary classification models, the ROC-AUC (Receiver Operating Characteristic - Area Under the Curve) score can be used to evaluate performance. It quantifies the ability of the model to distinguish between positive and negative instances across different decision thresholds. A higher ROC-AUC score (ranging from 0.5 to 1) indicates better performance, with 1 representing a perfect classifier and 0.5 indicating performance no better than random chance. There is a more detailed discussion here: https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc.

The metric can be calculated using the `roc_auc_score` function from the `sklearn.metrics` module, e.g.

```python
from sklearn.metrics import roc_auc_score

# Assuming you have true labels (y_true) and predicted probabilities (y_pred_prob)
y_true = [...]  
y_pred_prob = [...]  

# Calculate ROC-AUC
roc_auc = roc_auc_score(y_true, y_pred_prob)

# Display the result
print(f'ROC-AUC Score: {roc_auc:.4f}')
```
</details>

In [32]:
# Spare cell




## Model analysis and discussion

How well does your final model perform? Think of metrics and plots that are useful to dig a little deeper.

Compare against the best-performing model on the [MatBench](https://matbench.materialsproject.org) leaderboard.  With limited resources, don't expect to match this performance, but you should do better than a baseline model.

In [33]:
# Spare cell




## Large Language Model (LLM) usage declaration

Acknowledge use of a generative model during your assignment. Points to consider:

* State which LLM (e.g. GPT-4, Gemini, Co-Pilot)

* Specify tasks (e.g. summarising research or code snippets)

* Were any limitations/biases noted?

* How did you ensure ethical use?

In [34]:
# Spare cell




## ☘️ Final word

Good luck building your own model! We hope that you enjoyed the course and exercises. Dive deeper into the aspects that caught your interest. A useful starting point may be the [Resources](https://aronwalsh.github.io/MLforMaterials/Resources.html) page.

Remember that submission is on Blackboard and you should upload both the completed Juypter Notebook (`.ipynb` file), as well as your recorded narrated presentation (maximum 5 minutes; see guides on using [Zoom](https://www.youtube.com/watch?v=H9qhoAIzW3E) or [Powerpoint](https://www.youtube.com/watch?v=Y5dgwwa5XRA) for this purpose).

# Task
```python
from sklearn.base import BaseEstimator, TransformerMixin
import pandas as pd
from elementembeddings.composition import composition_featuriser
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
from sklearn.metrics import roc_auc_score
from scipy.stats import randint

# 1. Define Custom Featurizer Transformer
class CompositionFeaturizerTransformer(BaseEstimator, TransformerMixin):
    """
    A scikit-learn compatible transformer to apply composition featurization.
    """
    def __init__(self, embedding="magpie", stats=["mean"], composition_column="composition"):
        self.embedding = embedding
        self.stats = stats
        self.composition_column = composition_column
        self.featurizer_obj = None # To store the featurizer if stateful operations are needed

    def fit(self, X, y=None):
        # In this simple case, the featurizer does not need to be fitted to data.
        # It's a stateless transformation.
        return self

    def transform(self, X):
        """
        Featurizes compositions from a DataFrame.

        Args:
            X (pd.DataFrame or pd.Series): Input data containing the composition column.

        Returns:
            np.ndarray: Featurized data as a NumPy array.
        """
        if isinstance(X, pd.Series):
            compositions = X
        elif isinstance(X, pd.DataFrame):
            if self.composition_column not in X.columns:
                raise ValueError(f"Composition column '{self.composition_column}' not found in DataFrame.")
            compositions = X[self.composition_column]
        else:
            raise TypeError("Input X must be a pandas Series or DataFrame.")

        # Apply featurization
        featurized_df = composition_featuriser(
            compositions,
            embedding=self.embedding,
            stats=self.stats
        )

        # Drop the 'formula' column if it exists, as it's not a feature
        cols_to_drop = ['formula']
        final_features = featurized_df.drop(columns=cols_to_drop, errors='ignore')

        return final_features.values


# Assuming 'df' and 'y' are already loaded and preprocessed as per the notebook.
# df contains 'composition' and 'is_metal'
# y is the 'is_metal' target array (converted to int)

# Make a copy to avoid modifying the original DataFrame in place during transformation
X_data_for_pipeline = df[['composition']].copy()
y_data_for_pipeline = df['is_metal'].values

# 2. Set up Model Pipeline
# We'll use RandomForestClassifier for this example.
# The custom featurizer will output a numpy array, which the classifier can directly use.
pipeline = Pipeline([
    ('featurizer', CompositionFeaturizerTransformer(embedding="magpie", stats=["mean"])),
    ('classifier', RandomForestClassifier(random_state=CID))
])

# 3. Define Hyperparameter Search Space
# For RandomizedSearchCV, we define distributions for hyperparameters.
param_distributions = {
    # No featurizer hyperparameters to tune for now, but could add, e.g., 'featurizer__stats': [['mean'], ['maxpool']]
    'classifier__n_estimators': randint(50, 200),  # Number of trees in the forest
    'classifier__max_features': ['sqrt', 'log2', None], # Number of features to consider when looking for the best split
    'classifier__max_depth': randint(5, 30),        # Maximum depth of the tree
    'classifier__min_samples_split': randint(2, 20), # Minimum number of samples required to split an internal node
    'classifier__min_samples_leaf': randint(1, 20),  # Minimum number of samples required to be at a leaf node
    'classifier__criterion': ['gini', 'entropy']
}

# 4. Configure Inner and Outer Cross-Validation
# Using StratifiedKFold for classification to preserve class distribution.
# For Matbench, shuffle=False is often specified, but for general robust evaluation, shuffle=True is common.
# Here, we follow the general approach.
outer_cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=CID)
inner_cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=CID) # Typically fewer splits for inner CV to save computation

# Store ROC-AUC scores from outer folds
outer_roc_auc_scores = []

# 5. Implement Nested Cross-Validation Loop
print("Starting Nested Cross-Validation...")
for i, (train_index, test_index) in enumerate(outer_cv.split(X_data_for_pipeline, y_data_for_pipeline)):
    print(f"\n--- Outer Fold {i+1}/{outer_cv.get_n_splits()} ---")

    # Split data for outer loop
    X_outer_train, X_outer_test = X_data_for_pipeline.iloc[train_index], X_data_for_pipeline.iloc[test_index]
    y_outer_train, y_outer_test = y_data_for_pipeline[train_index], y_data_for_pipeline[test_index]

    # Set up RandomizedSearchCV for the inner loop
    # n_iter controls the number of parameter settings that are sampled.
    # More iterations mean a more exhaustive search, but also longer runtime.
    random_search = RandomizedSearchCV(
        estimator=pipeline,
        param_distributions=param_distributions,
        n_iter=50,  # Number of parameter settings that are sampled. Adjust as needed.
        cv=inner_cv,
        scoring='roc_auc',
        random_state=CID,
        n_jobs=-1, # Use all available cores
        verbose=0 # Set to 1 or 2 for more detailed output during search
    )

    # Perform hyperparameter optimization on the outer training set
    random_search.fit(X_outer_train, y_outer_train)

    # Get the best estimator from the inner search
    best_model = random_search.best_estimator_
    print(f"Best hyperparameters for Outer Fold {i+1}: {random_search.best_params_}")

    # Evaluate the best model on the outer test set
    y_pred_proba = best_model.predict_proba(X_outer_test)[:, 1] # Get probability of positive class
    fold_roc_auc = roc_auc_score(y_outer_test, y_pred_proba)
    outer_roc_auc_scores.append(fold_roc_auc)
    print(f"ROC-AUC score for Outer Fold {i+1}: {fold_roc_auc:.4f}")

# 6. Report Nested CV Results
mean_roc_auc = np.mean(outer_roc_auc_scores)
std_roc_auc = np.std(outer_roc_auc_scores)

print("\n--- Nested Cross-Validation Results ---")
print(f"Individual ROC-AUC scores for outer folds: {outer_roc_auc_scores}")
print(f"Mean ROC-AUC: {mean_roc_auc:.4f}")
print(f"Standard Deviation of ROC-AUC: {std_roc_auc:.4f}")
```

## Define Custom Featurizer Transformer

### Subtask:
Create a scikit-learn compatible transformer class that wraps the `composition_featuriser` function. This will ensure featurization is applied correctly within the pipeline to avoid data leakage.


**Reasoning**:
The subtask requires defining a custom scikit-learn compatible transformer. This code block will import the necessary libraries and define the `CompositionFeaturizerTransformer` class as per the instructions, including the `__init__`, `fit`, and `transform` methods.



In [47]:
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from elementembeddings.composition import composition_featuriser

class CompositionFeaturizerTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, embedding="magpie", stats=["mean"], composition_column="composition"):
        self.embedding = embedding
        self.stats = stats
        self.composition_column = composition_column

    def fit(self, X, y=None):
        # Featurizer is stateless, so no fitting is needed
        return self

    def transform(self, X):
        if isinstance(X, pd.Series):
            compositions = X
        elif isinstance(X, pd.DataFrame):
            if self.composition_column not in X.columns:
                raise ValueError(f"Composition column '{self.composition_column}' not found in DataFrame.")
            compositions = X[self.composition_column]
        else:
            raise TypeError("Input X must be a pandas Series or DataFrame.")

        # Call composition_featuriser to get features
        featurized_df = composition_featuriser(compositions, embedding=self.embedding, stats=self.stats)

        # Drop the 'formula' column if it exists, as it's not a feature
        if 'formula' in featurized_df.columns:
            featurized_df = featurized_df.drop(columns=['formula'])

        return featurized_df.values

print("CompositionFeaturizerTransformer class defined.")

CompositionFeaturizerTransformer class defined.


## Set up Model Pipeline

### Subtask:
Construct a scikit-learn pipeline that includes the custom featurizer transformer and a machine learning classifier (e.g., RandomForestClassifier).


**Reasoning**:
The subtask requires constructing a scikit-learn pipeline. This code block will prepare the input data, import necessary classes, instantiate the custom featurizer and a Random Forest Classifier, and then combine them into a pipeline as instructed.



In [48]:
import pandas as pd
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier

# 1. Make a copy of the 'composition' column from the DataFrame `df`
X_data_for_pipeline = df[['composition']].copy()

# 2. Extract the 'is_metal' column from `df` and convert it to a NumPy array of integers
y_data_for_pipeline = df['is_metal'].astype(int).values

# 4. Instantiate the CompositionFeaturizerTransformer
# The CompositionFeaturizerTransformer class is already defined in a previous step.
featurizer_instance = CompositionFeaturizerTransformer(embedding="magpie", stats=["mean"], composition_column="composition")

# 5. Instantiate a RandomForestClassifier
classifier_instance = RandomForestClassifier(random_state=CID)

# 6. Create a scikit-learn Pipeline
pipeline = Pipeline([
    ('featurizer', featurizer_instance),
    ('classifier', classifier_instance)
])

print("Pipeline constructed successfully:")
print(pipeline)

Pipeline constructed successfully:
Pipeline(steps=[('featurizer', CompositionFeaturizerTransformer()),
                ('classifier', RandomForestClassifier(random_state=2198282))])


## Define Hyperparameter Search Space

### Subtask:
Specify the hyperparameter distributions for the `RandomizedSearchCV` within the pipeline. This will include parameters for the featurizer (if any) and the classifier.


**Reasoning**:
To define the hyperparameter search space for RandomizedSearchCV, I need to import `randint` and then create the `param_distributions` dictionary with the specified distributions for the Random Forest Classifier's hyperparameters.



In [49]:
from scipy.stats import randint

# Define Hyperparameter Search Space
param_distributions = {
    'classifier__n_estimators': randint(50, 200),  # Number of trees in the forest
    'classifier__max_features': ['sqrt', 'log2', None], # Number of features to consider when looking for the best split
    'classifier__max_depth': randint(5, 30),        # Maximum depth of the tree
    'classifier__min_samples_split': randint(2, 20), # Minimum number of samples required to split an internal node
    'classifier__min_samples_leaf': randint(1, 20),  # Minimum number of samples required to be at a leaf node
    'classifier__criterion': ['gini', 'entropy']
}

print("Hyperparameter search space 'param_distributions' defined.")

Hyperparameter search space 'param_distributions' defined.


**Reasoning**:
The next step is to configure the inner and outer cross-validation strategies, using `StratifiedKFold` to ensure class distribution is preserved during splitting, which is important for classification tasks. I will then instantiate `RandomizedSearchCV` for the inner loop.



In [50]:
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold

# Configure Inner and Outer Cross-Validation
# Using StratifiedKFold for classification to preserve class distribution.
# For Matbench, shuffle=False is often specified, but for general robust evaluation, shuffle=True is common.
# Here, we follow the general approach.
outer_cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=CID)
inner_cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=CID) # Typically fewer splits for inner CV to save computation

# Set up RandomizedSearchCV for the inner loop
# n_iter controls the number of parameter settings that are sampled.
# More iterations mean a more exhaustive search, but also longer runtime.
random_search_inner = RandomizedSearchCV(
    estimator=pipeline,
    param_distributions=param_distributions,
    n_iter=50,  # Number of parameter settings that are sampled. Adjust as needed.
    cv=inner_cv,
    scoring='roc_auc',
    random_state=CID,
    n_jobs=-1, # Use all available cores
    verbose=0 # Set to 1 or 2 for more detailed output during search
)

print("Cross-validation objects and RandomizedSearchCV for inner loop configured.")

Cross-validation objects and RandomizedSearchCV for inner loop configured.


**Reasoning**:
The next step is to implement the nested cross-validation loop by iterating through the outer folds, performing the hyperparameter search on the inner folds using `random_search_inner`, evaluating the best model on the outer test set, and collecting the ROC-AUC scores.



In [51]:
from sklearn.metrics import roc_auc_score
import numpy as np

# Store ROC-AUC scores from outer folds
outer_roc_auc_scores = []

# Implement Nested Cross-Validation Loop
print("Starting Nested Cross-Validation...")
for i, (train_index, test_index) in enumerate(outer_cv.split(X_data_for_pipeline, y_data_for_pipeline)):
    print(f"\n--- Outer Fold {i+1}/{outer_cv.get_n_splits()} ---")

    # Split data for outer loop
    X_outer_train, X_outer_test = X_data_for_pipeline.iloc[train_index], X_data_for_pipeline.iloc[test_index]
    y_outer_train, y_outer_test = y_data_for_pipeline[train_index], y_data_for_pipeline[test_index]

    # Perform hyperparameter optimization on the outer training set
    # Use the pre-configured random_search_inner object
    random_search_inner.fit(X_outer_train, y_outer_train)

    # Get the best estimator from the inner search
    best_model = random_search_inner.best_estimator_
    print(f"Best hyperparameters for Outer Fold {i+1}: {random_search_inner.best_params_}")

    # Evaluate the best model on the outer test set
    y_pred_proba = best_model.predict_proba(X_outer_test)[:, 1] # Get probability of positive class
    fold_roc_auc = roc_auc_score(y_outer_test, y_pred_proba)
    outer_roc_auc_scores.append(fold_roc_auc)
    print(f"ROC-AUC score for Outer Fold {i+1}: {fold_roc_auc:.4f}")

# Report Nested CV Results
mean_roc_auc = np.mean(outer_roc_auc_scores)
std_roc_auc = np.std(outer_roc_auc_scores)

print("\n--- Nested Cross-Validation Results ---")
print(f"Individual ROC-AUC scores for outer folds: {outer_roc_auc_scores}")
print(f"Mean ROC-AUC: {mean_roc_auc:.4f}")
print(f"Standard Deviation of ROC-AUC: {std_roc_auc:.4f}")

Starting Nested Cross-Validation...

--- Outer Fold 1/5 ---


ValueError: 
All the 150 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
50 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
    self._final_estimator.fit(Xt, y, **last_step_params["fit"])
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_forest.py", line 360, in fit
    X, y = validate_data(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 2961, in validate_data
    X, y = check_X_y(X, y, **check_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 1389, in check_X_y
    check_consistent_length(X, y)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 475, in check_consistent_length
    raise ValueError(
ValueError: Found input variables with inconsistent numbers of samples: [3848, 2624]

--------------------------------------------------------------------------------
50 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
    self._final_estimator.fit(Xt, y, **last_step_params["fit"])
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_forest.py", line 360, in fit
    X, y = validate_data(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 2961, in validate_data
    X, y = check_X_y(X, y, **check_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 1389, in check_X_y
    check_consistent_length(X, y)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 475, in check_consistent_length
    raise ValueError(
ValueError: Found input variables with inconsistent numbers of samples: [3835, 2624]

--------------------------------------------------------------------------------
50 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
    self._final_estimator.fit(Xt, y, **last_step_params["fit"])
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_forest.py", line 360, in fit
    X, y = validate_data(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 2961, in validate_data
    X, y = check_X_y(X, y, **check_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 1389, in check_X_y
    check_consistent_length(X, y)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py", line 475, in check_consistent_length
    raise ValueError(
ValueError: Found input variables with inconsistent numbers of samples: [3851, 2624]


**Reasoning**:
The error "Found input variables with inconsistent numbers of samples" during nested cross-validation indicates a mismatch in the number of rows between the featurized input (X) and the target variable (y). This is likely caused by either the `composition_featuriser` implicitly dropping problematic compositions, or `sklearn`'s internal validation implicitly removing rows with `NaN` values, leading to a length discrepancy. To fix this, the `CompositionFeaturizerTransformer` needs to explicitly reindex the featurized output to match the input length, filling missing rows with NaNs, and a `SimpleImputer` must be added to the pipeline to handle these NaNs before the classifier receives the data.



In [52]:
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from elementembeddings.composition import composition_featuriser
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import SimpleImputer # Import SimpleImputer
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
from scipy.stats import randint
from sklearn.metrics import roc_auc_score

# Re-define CompositionFeaturizerTransformer with reindexing for robustness
class CompositionFeaturizerTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, embedding="magpie", stats=["mean"], composition_column="composition"):
        self.embedding = embedding
        self.stats = stats
        self.composition_column = composition_column

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        if isinstance(X, pd.Series):
            compositions = X
        elif isinstance(X, pd.DataFrame):
            if self.composition_column not in X.columns:
                raise ValueError(f"Composition column '{self.composition_column}' not found in DataFrame.")
            compositions = X[self.composition_column]
        else:
            raise TypeError("Input X must be a pandas Series or DataFrame.")

        original_index = compositions.index

        # Apply featurization
        featurized_df = composition_featuriser(
            compositions,
            embedding=self.embedding,
            stats=self.stats
        )

        # Drop the 'formula' column if it exists, as it's not a feature
        cols_to_drop_from_features = ['formula'] # Use a different variable name to avoid conflict
        final_features = featurized_df.drop(columns=cols_to_drop_from_features, errors='ignore')

        # Reindex to ensure consistent length with the original input compositions
        # This will fill with NaNs if any rows were unexpectedly dropped by the featurizer
        final_features = final_features.reindex(original_index)

        return final_features.values

# Assuming X_data_for_pipeline, y_data_for_pipeline are already defined from previous steps
# And CID, outer_cv, inner_cv, param_distributions are also defined.

# Re-define the pipeline to include the SimpleImputer
pipeline = Pipeline([
    ('featurizer', CompositionFeaturizerTransformer(embedding="magpie", stats=["mean"], composition_column="composition")),
    ('imputer', SimpleImputer(strategy='mean')), # Impute NaNs after featurization
    ('classifier', RandomForestClassifier(random_state=CID))
])

# Re-define random_search_inner with the updated pipeline
random_search_inner = RandomizedSearchCV(
    estimator=pipeline,
    param_distributions=param_distributions,
    n_iter=50,
    cv=inner_cv,
    scoring='roc_auc',
    random_state=CID,
    n_jobs=-1,
    verbose=0
)

# Store ROC-AUC scores from outer folds
outer_roc_auc_scores = []

# Implement Nested Cross-Validation Loop
print("Starting Nested Cross-Validation...")
for i, (train_index, test_index) in enumerate(outer_cv.split(X_data_for_pipeline, y_data_for_pipeline)):
    print(f"\n--- Outer Fold {i+1}/{outer_cv.get_n_splits()} ---")

    # Split data for outer loop
    X_outer_train, X_outer_test = X_data_for_pipeline.iloc[train_index], X_data_for_pipeline.iloc[test_index]
    y_outer_train, y_outer_test = y_data_for_pipeline[train_index], y_data_for_pipeline[test_index]

    # Perform hyperparameter optimization on the outer training set
    random_search_inner.fit(X_outer_train, y_outer_train)

    # Get the best estimator from the inner search
    best_model = random_search_inner.best_estimator_
    print(f"Best hyperparameters for Outer Fold {i+1}: {random_search_inner.best_params_}")

    # Evaluate the best model on the outer test set
    y_pred_proba = best_model.predict_proba(X_outer_test)[:, 1] # Get probability of positive class
    fold_roc_auc = roc_auc_score(y_outer_test, y_pred_proba)
    outer_roc_auc_scores.append(fold_roc_auc)
    print(f"ROC-AUC score for Outer Fold {i+1}: {fold_roc_auc:.4f}")

# Report Nested CV Results
mean_roc_auc = np.mean(outer_roc_auc_scores)
std_roc_auc = np.std(outer_roc_auc_scores)

print("\n--- Nested Cross-Validation Results ---")
print(f"Individual ROC-AUC scores for outer folds: {outer_roc_auc_scores}")
print(f"Mean ROC-AUC: {mean_roc_auc:.4f}")
print(f"Standard Deviation of ROC-AUC: {std_roc_auc:.4f}")


ImportError: cannot import name 'SimpleImputer' from 'sklearn.preprocessing' (/usr/local/lib/python3.12/dist-packages/sklearn/preprocessing/__init__.py)

**Reasoning**:
The previous code failed because `SimpleImputer` is imported from the wrong module. It should be imported from `sklearn.impute`, not `sklearn.preprocessing`.



In [53]:
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from elementembeddings.composition import composition_featuriser
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer # Corrected import for SimpleImputer
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
from scipy.stats import randint
from sklearn.metrics import roc_auc_score

# Re-define CompositionFeaturizerTransformer with reindexing for robustness
class CompositionFeaturizerTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, embedding="magpie", stats=["mean"], composition_column="composition"):
        self.embedding = embedding
        self.stats = stats
        self.composition_column = composition_column

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        if isinstance(X, pd.Series):
            compositions = X
        elif isinstance(X, pd.DataFrame):
            if self.composition_column not in X.columns:
                raise ValueError(f"Composition column '{self.composition_column}' not found in DataFrame.")
            compositions = X[self.composition_column]
        else:
            raise TypeError("Input X must be a pandas Series or DataFrame.")

        original_index = compositions.index

        # Apply featurization
        featurized_df = composition_featuriser(
            compositions,
            embedding=self.embedding,
            stats=self.stats
        )

        # Drop the 'formula' column if it exists, as it's not a feature
        cols_to_drop_from_features = ['formula'] # Use a different variable name to avoid conflict
        final_features = featurized_df.drop(columns=cols_to_drop_from_features, errors='ignore')

        # Reindex to ensure consistent length with the original input compositions
        # This will fill with NaNs if any rows were unexpectedly dropped by the featurizer
        final_features = final_features.reindex(original_index)

        return final_features.values

# Assuming X_data_for_pipeline, y_data_for_pipeline are already defined from previous steps
# And CID, outer_cv, inner_cv, param_distributions are also defined.

# Re-define the pipeline to include the SimpleImputer
pipeline = Pipeline([
    ('featurizer', CompositionFeaturizerTransformer(embedding="magpie", stats=["mean"], composition_column="composition")),
    ('imputer', SimpleImputer(strategy='mean')), # Impute NaNs after featurization
    ('classifier', RandomForestClassifier(random_state=CID))
])

# Re-define random_search_inner with the updated pipeline
random_search_inner = RandomizedSearchCV(
    estimator=pipeline,
    param_distributions=param_distributions,
    n_iter=50,
    cv=inner_cv,
    scoring='roc_auc',
    random_state=CID,
    n_jobs=-1,
    verbose=0
)

# Store ROC-AUC scores from outer folds
outer_roc_auc_scores = []

# Implement Nested Cross-Validation Loop
print("Starting Nested Cross-Validation...")
for i, (train_index, test_index) in enumerate(outer_cv.split(X_data_for_pipeline, y_data_for_pipeline)):
    print(f"\n--- Outer Fold {i+1}/{outer_cv.get_n_splits()} ---")

    # Split data for outer loop
    X_outer_train, X_outer_test = X_data_for_pipeline.iloc[train_index], X_data_for_pipeline.iloc[test_index]
    y_outer_train, y_outer_test = y_data_for_pipeline[train_index], y_data_for_pipeline[test_index]

    # Perform hyperparameter optimization on the outer training set
    random_search_inner.fit(X_outer_train, y_outer_train)

    # Get the best estimator from the inner search
    best_model = random_search_inner.best_estimator_
    print(f"Best hyperparameters for Outer Fold {i+1}: {random_search_inner.best_params_}")

    # Evaluate the best model on the outer test set
    y_pred_proba = best_model.predict_proba(X_outer_test)[:, 1] # Get probability of positive class
    fold_roc_auc = roc_auc_score(y_outer_test, y_pred_proba)
    outer_roc_auc_scores.append(fold_roc_auc)
    print(f"ROC-AUC score for Outer Fold {i+1}: {fold_roc_auc:.4f}")

# Report Nested CV Results
mean_roc_auc = np.mean(outer_roc_auc_scores)
std_roc_auc = np.std(outer_roc_auc_scores)

print("\n--- Nested Cross-Validation Results ---")
print(f"Individual ROC-AUC scores for outer folds: {outer_roc_auc_scores}")
print(f"Mean ROC-AUC: {mean_roc_auc:.4f}")
print(f"Standard Deviation of ROC-AUC: {std_roc_auc:.4f}")

Starting Nested Cross-Validation...

--- Outer Fold 1/5 ---
Featurising compositions...


100%|██████████| 3936/3936 [00:10<00:00, 388.75it/s]


Computing feature vectors...


100%|██████████| 3936/3936 [00:00<00:00, 165677.62it/s]


Best hyperparameters for Outer Fold 1: {'classifier__criterion': 'entropy', 'classifier__max_depth': 14, 'classifier__max_features': 'log2', 'classifier__min_samples_leaf': 19, 'classifier__min_samples_split': 10, 'classifier__n_estimators': 138}
Featurising compositions...


100%|██████████| 985/985 [00:02<00:00, 339.31it/s]


Computing feature vectors...


100%|██████████| 985/985 [00:00<00:00, 81355.39it/s]


ROC-AUC score for Outer Fold 1: 0.5242

--- Outer Fold 2/5 ---


KeyboardInterrupt: 

**Reasoning**:
The previous execution indicated that the nested cross-validation process was started, but then failed with a `ValueError` related to inconsistent sample sizes, followed by an `ImportError` for `SimpleImputer`. The code block provided previously included the fix for the `ImportError` by correctly importing `SimpleImputer` from `sklearn.impute` and a logic to reindex and impute in the `CompositionFeaturizerTransformer`. I need to execute this corrected code block to see if the nested cross-validation runs successfully and provides the expected results.



In [ ]:
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from elementembeddings.composition import composition_featuriser
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer # Corrected import for SimpleImputer
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
from scipy.stats import randint
from sklearn.metrics import roc_auc_score

# Re-define CompositionFeaturizerTransformer with reindexing for robustness
class CompositionFeaturizerTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, embedding="magpie", stats=["mean"], composition_column="composition"):
        self.embedding = embedding
        self.stats = stats
        self.composition_column = composition_column

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        if isinstance(X, pd.Series):
            compositions = X
        elif isinstance(X, pd.DataFrame):
            if self.composition_column not in X.columns:
                raise ValueError(f"Composition column '{self.composition_column}' not found in DataFrame.")
            compositions = X[self.composition_column]
        else:
            raise TypeError("Input X must be a pandas Series or DataFrame.")

        original_index = compositions.index

        # Apply featurization
        featurized_df = composition_featuriser(
            compositions,
            embedding=self.embedding,
            stats=self.stats
        )

        # Drop the 'formula' column if it exists, as it's not a feature
        cols_to_drop_from_features = ['formula'] # Use a different variable name to avoid conflict
        final_features = featurized_df.drop(columns=cols_to_drop_from_features, errors='ignore')

        # Reindex to ensure consistent length with the original input compositions
        # This will fill with NaNs if any rows were unexpectedly dropped by the featurizer
        final_features = final_features.reindex(original_index)

        return final_features.values

# Assuming X_data_for_pipeline, y_data_for_pipeline are already defined from previous steps
# And CID, outer_cv, inner_cv, param_distributions are also defined.

# Re-define the pipeline to include the SimpleImputer
pipeline = Pipeline([
    ('featurizer', CompositionFeaturizerTransformer(embedding="magpie", stats=["mean"], composition_column="composition")),
    ('imputer', SimpleImputer(strategy='mean')), # Impute NaNs after featurization
    ('classifier', RandomForestClassifier(random_state=CID))
])

# Re-define random_search_inner with the updated pipeline
random_search_inner = RandomizedSearchCV(
    estimator=pipeline,
    param_distributions=param_distributions,
    n_iter=50,
    cv=inner_cv,
    scoring='roc_auc',
    random_state=CID,
    n_jobs=-1,
    verbose=0
)

# Store ROC-AUC scores from outer folds
outer_roc_auc_scores = []

# Implement Nested Cross-Validation Loop
print("Starting Nested Cross-Validation...")
for i, (train_index, test_index) in enumerate(outer_cv.split(X_data_for_pipeline, y_data_for_pipeline)):
    print(f"\n--- Outer Fold {i+1}/{outer_cv.get_n_splits()} ---")

    # Split data for outer loop
    X_outer_train, X_outer_test = X_data_for_pipeline.iloc[train_index], X_data_for_pipeline.iloc[test_index]
    y_outer_train, y_outer_test = y_data_for_pipeline[train_index], y_data_for_pipeline[test_index]

    # Perform hyperparameter optimization on the outer training set
    random_search_inner.fit(X_outer_train, y_outer_train)

    # Get the best estimator from the inner search
    best_model = random_search_inner.best_estimator_
    print(f"Best hyperparameters for Outer Fold {i+1}: {random_search_inner.best_params_}")

    # Evaluate the best model on the outer test set
    y_pred_proba = best_model.predict_proba(X_outer_test)[:, 1] # Get probability of positive class
    fold_roc_auc = roc_auc_score(y_outer_test, y_pred_proba)
    outer_roc_auc_scores.append(fold_roc_auc)
    print(f"ROC-AUC score for Outer Fold {i+1}: {fold_roc_auc:.4f}")

# Report Nested CV Results
mean_roc_auc = np.mean(outer_roc_auc_scores)
std_roc_auc = np.std(outer_roc_auc_scores)

print("\n--- Nested Cross-Validation Results ---")
print(f"Individual ROC-AUC scores for outer folds: {outer_roc_auc_scores}")
print(f"Mean ROC-AUC: {mean_roc_auc:.4f}")
print(f"Standard Deviation of ROC-AUC: {std_roc_auc:.4f}")


Starting Nested Cross-Validation...

--- Outer Fold 1/5 ---
